libsel4/include/sel4/constants.h

```diff
@@ -52,7 +52,16 @@ enum priorityConstants {
 enum seL4_MsgLimits {
     seL4_MsgLengthBits = 7,
     seL4_MsgExtraCapBits = 2,
+#if defined(CONFIG_HAVE_CHERI)
+    /* As CHERI doubles the register size, this affects the size of msg[]
+     * array, so we halve it here to have an overall fixed-size IPC buffer
+     * for both CHERI and non-CHERI builds. We further substract 1 more CHERI
+     * word to meet the alignment constraint set on the IPC buffer struct.
+     */
+    seL4_MsgMaxLength = 59
+#else
     seL4_MsgMaxLength = 120
+#endif
```

Comment:
> > [cheri][ipc] Make IPC's msg[] seL4_Register type
> 
> > Hence, it is a deliberate decision to have msg[] the same as
> > rword_t/seL4_Register.
> 
> This decision is something that should have more discussion and I think the design choice here likely needs to be presented at the level of the RFC.
> 

I am happy to discuss that, I'll just quote my reasons in the commit message here again:

> Enable seL4's IPC buffer to seemingly hold CHERI-sized words.
> Message registers in the IPC buffer intuitively correspond to the
> underlying HW register/word's size/format, not an integer or a pointer.
> It is used during IPC for various use cases such as:
> 1) Transferring HW registers (eg read/write/copy TCBRegisterContext)
> 2) Transferring register context to a fault handler
> 3) Transferring HW-register-sized data between user threads
> 4) System call arguments (eg TLSBase, VCPU reg)

> Currently the IPC buffer is defined in terms of 120 words for all architectures where a word is the same size for a pointer and for data. A benefit of this choice is the direct mapping of msg registers onto hardware registers for fast IPC transfer.
> 
> I argue that `seL4_Word msg[seL4_MsgMaxLength];` should be preserved on CHERI configurations to continue supporting programs and interfaces that make use the IPC buffer to transfer messages. When transferring values that aren't CHERI pointers (The common case), halving the number of message registers and doubling their size would add a bunch of porting effort and make it harder to write programs that target both CHERI and non-CHERI hardware. We should consider it's possible to avoid this.
> 

The common case is transferring untagged data "words" or messages, not 64-bit integers. `seL4_Word` is no longer representing a HW word here, it's just an integer. The name is just left as it is (probably misleadingly) to avoid more disruptive changes, but it's documented in the commit message and it should only be used to hold integers. 

I believe that this change has no effect at all on continuing supporting programs and interfaces, on the contrary, it's necessary to:
1) naturally respect, align with, and intentionally support the purecap ABI (where your words are really no longer `seL4_Word` integers), and,
2) avoid extra porting and maintenance headache (e.g., casting, unions, ifdefs, etc). This claim is based on my experimentations with sel4test and sel4bench that worked just fine with that change whether it was 120 words, or 59 words, with no changes at all this part, whether for CHERI or non-CHERI builds. 
3) Perform 2x faster IPC (e.g., doing 128-bit untagged data copies, instead of 64-bit), when transferring fixed message sizes. 

I don't mind preserving the `seL4_MsgMaxLength` to be 120 as it is (and I actually kept it that way in the previous PR), but that will require increasing the IPC buffer size, and I got the impression you don't want that from the discussion on the previous PR, so I changed it here. 

> There are other options we could consider such as defining an alternate `seL4_IPCBuffer` type that can be used for invocations that transfer hardware caps values. Or inserting a `union` into the storage occupied by the `msg` field that allows it to be used through a `seL4_Word` interface or instead through a `seL4_HWCap` interface. (I think it would be worth considering aligning up the `msg` field to 128bits in this scenario).

Reply:
Like I said before, I would prohibit any tagged pointer transfer via memory across tasks and address spaces.

I would limit tagged pointer passing only via the register arguments of a select few syscalls, and add new single register read/write ones to avoid any need for tagged pointer passing via the IPC buffer.

seL4 can disallow multiple mappings of tagged pointer pages, as enabled with the new CHERI PTE flag set by:
- Only allowing a tagged pointer memory mapping if the page cap used is the only copy of that cap. This can be done by checking whether the parent object is an UT and whether there are no siblings.
- Prohibit cap copying of a page cap which is mapped with the CHERI PTE flag.
You need the flag, you really don't want to enable tagged pointers for shared memory.

The above may result in a slightly bigger diff, but the actual behavioural and semantic change would be much smaller.

Reply:
Hi Indan,

> Like I said before, I would prohibit any tagged pointer transfer via memory across tasks and address spaces.
> 
Yeah we agree on the across address spaces part. 

> I would limit tagged pointer passing only via the register arguments of a select few syscalls, and add new single register read/write ones to avoid any need for tagged pointer passing via the IPC buffer.
> 
I have been thinking about this approach as well; do you think if the user maps the IPC buffer page without CHERI PTE capability load/store permissions, this will yield to that limiting behaviour? This way, only the fast message registers will be able to contain tagged pointers, and those get saved/restored from the TCB's archRegisters, not the IPC buffer.

> seL4 can disallow multiple mappings of tagged pointer pages, as enabled with the new CHERI PTE flag set by:
> 
> * Only allowing a tagged pointer memory mapping if the page cap used is the only copy of that cap. This can be done by checking whether the parent object is an UT and whether there are no siblings.
> * Prohibit cap copying of a page cap which is mapped with the CHERI PTE flag.
>   You need the flag, you really don't want to enable tagged pointers for shared memory.
> 
Exactly, we generally disallow tagged pointers for shared memory in CheriBSD (eg mmap with `MAP_SHARED`). 

> The above may result in a slightly bigger diff, but the actual behavioural and semantic change would be much smaller.
I will think about this further.

Reply:
> Like I said before, I would prohibit any tagged pointer transfer via memory across tasks and address spaces.

I made a separate thread for having that discussion: https://github.com/seL4/seL4/pull/1344#discussion_r1835844064

I'd like for this thread to be about whether the IPC buffer storage should change it's type. 

> `seL4_Word` is no longer representing a HW word here, it's just an integer. The name is just left as it is (probably misleadingly) to avoid more disruptive changes, but it's documented in the commit message and it should only be used to hold integers.

I don't think we agree on what a HW word is. Usually [1], it's the natural unit of **data** used by the processor which is still 64 bits in an ARM C64 architecture because the natural size of data used by instructions is still 64bits. I can see how a disagreement in this definition leads to downstream disagreements about what the natural size of base types should be things like the IPC buffer.


[1]: https://en.wikipedia.org/wiki/Word_(computer_architecture)

Reply:
I've been thinking about this more, and while I think that `seL4_Word` is correct to stay at the machine word size of 64 bits, it also makes sense that `seL4_Word` isn't appropriate for the msg register type on a CHERI system, and a better type would be `seL4_Register` as proposed in the commit. 

I'm still unsure about making this change universal across non-CHERI architectures, but introducing a new type `seL4_Register` for the register size should be considered?

Reply:
There are two options here 

1. Change seL4_Word (and word_t) to always be of a machine word_size, and just use `[u]int[xlen]_t` types for anything used for just integers. Even though I think this is the right clean approach, but this will be significantly disruptive, and I have actually done it.
2. Keep `seL4_Word` but emphasise that in CHERI systems, this is purely an integer. Introduce `seL4_Register` to be used *only* for CHERI.
3. Follow 2), while trying to be more intentional and use seL4_Register in shared code (non-CHERI and CHERI) to differentiate between using integers, pointers, messages registers, HW registers, etc. 

The advantage of 3 is that we would very much like to reduce `#ifdef CHERI #else` etc and avoid code duplications as possible. This will reduce maintenance effort for someone who doesn't know about CHERI, but is changing something in the non-CHERI `#ifdef` code that might break CHERI. Let alone the fact that it's more intentional/readable when it comes to namings and encourages good coding practices. That's why I went of 3 to meet halfway there.  

> I've been thinking about this more, and while I think that `seL4_Word` is correct to stay at the machine word size of 64 bits, it also makes sense that `seL4_Word` isn't appropriate for the msg register type on a CHERI system [1], and a better type would be `seL4_Register` as proposed in the commit.
> 
> I'm still unsure about making this change universal across non-CHERI architectures, but introducing a new type `seL4_Register` for the register size should be considered?
> 
Yeah that's what I did in this PR? 

There are three options here: 

1. Change `seL4_Word` (and `word_t`) to always be of a machine word/register size, and just use `[u]int[xlen]_t` types for anything used for just integers. Even though I think this is the right clean approach, this will be significantly disruptive, and I have actually done it, but won't want to scare you off even more :)
2. Keep `seL4_Word` but emphasise that in CHERI systems, this is purely an unsigned integer. Introduce `seL4_Register` to be used *only* for CHERI.
3. Follow 2), while trying to be more intentional and use `seL4_Register` in shared code (non-CHERI and CHERI) to differentiate between using integers, pointers, messages registers, HW registers, etc. 

The advantage of 3 is that we would very much like to reduce `#ifdef CHERI #else` etc and avoid code duplications as possible. This will reduce maintenance effort for someone who doesn't know about CHERI, but is changing something in the non-CHERI `#ifdef` code that might break CHERI. Let alone the fact that it's more intentional/readable when it comes to namings and encourages good coding practices, even for non-CHERI seL4. That's why I went of 3 to meet halfway there. I do understand this might have some verification effects in non-CHERI builds. This was expected to exist but be minimal; we aren't trying to completely have 0 verification effects. But, IMO, that's quite minimal (eg the commit I mentioned), whether it's just type renamings of types that always lower to `unsigned long` or some function prototypes, and finally accessing a few fields in the IPC buffer (eg `caps_or_badges`, which I am fine to ifdef). The question is, are these verification-visible changes tolerable and the verification tools can easily handle such minimal changes (from system's perspective)? 

> [1]: Because of how the kernel treats the first 4 msg registers as always belonging in a hardware register for all IPC transfers and for how the other fields are accessed by the kernel for IPC transfer in `copyMRs`.

Reply:
> I don't mind preserving the `seL4_MsgMaxLength` to be 120 as it is (and I actually kept it that way in the previous PR), but that will require increasing the IPC buffer size, and I got the impression you don't want that from the discussion on the previous PR, so I changed it here.

@kent-mcleod Any preference on that change? I am happy to do either

Reply:
> do you think if the user maps the IPC buffer page without CHERI PTE capability load/store permissions, this will yield to that limiting behaviour?

It is the kernel doing IPC buffer data copies without preserving tags that limits the behaviour. Then it doesn't matter what the PTE flag is for the page containing the IPC buffer.

(f the kernel uses it's own mapping to do the copying, it ignores the user's flag anyway. I don't want the tagged pointer PTE flag set on any kernel page mapping. However, to save and restore user context it does need to set it (but maybe it can be limited to load and store only).

> This way, only the fast message registers will be able to contain tagged pointers, and those get saved/restored from the TCB's archRegisters, not the IPC buffer.

Exactly. But only the few CHERI-aware syscalls will retain the tag, others will just store it as a word_t, probably avoiding the tagged load. I'd probably keep `getSyscallArg` as it is and add a CHERI-aware version for the special syscalls to use, to make it explicit.

> I'd like for this thread to be about whether the IPC buffer storage should change it's type.

If no tagged pointers are passed via the IPC buffer (because that would be passing CHERI pointers via memory), its type stays the same. That's why I replied here.

Reply:
> > do you think if the user maps the IPC buffer page without CHERI PTE capability load/store permissions, this will yield to that limiting behaviour?
> 
> It is the kernel doing IPC buffer data copies without preserving tags that limits the behaviour. Then it doesn't matter what the PTE flag is for the page containing the IPC buffer.
> 
I thought the main reason for doing so is to give the user the option to select a security policy that prevents the IPC buffer to contain tagged pointers, and hence, control tagged CHERI capability propagation at *user-level* during IPC. 

> (f the kernel uses it's own mapping to do the copying, it ignores the user's flag anyway. I don't want the tagged pointer PTE flag set on any kernel page mapping. However, to save and restore user context it does need to set it (but maybe it can be limited to load and store only).
> 
Yeah the kernel would ignore the user's for its own mappings, but why does it matter at all if the kernel is able to save capabilities in its mappings as long as the user has the ability to control capability propagation (to allow it or not) during IPC?

Please note that we still want to give the user the option to pass capabilities in the IPC buffer if they need/want to, whether it's inter-AS or intra-AS (I thought we agreed on that?), and not completely prevent that by hardcoding this policy in the kernel. 

> > This way, only the fast message registers will be able to contain tagged pointers, and those get saved/restored from the TCB's archRegisters, not the IPC buffer.
> 
> Exactly. But only the few CHERI-aware syscalls will retain the tag, others will just store it as a word_t, probably avoiding the tagged load. I'd probably keep `getSyscallArg` as it is and add a CHERI-aware version for the special syscalls to use, to make it explicit.
> 

OK so we agree that it's possible for the user to limit tagged capability propagation in the IPC buffer itself by clearing the PTE bits.

Retaining the tag or not is better handled by explicitly "clearing" the tag bit rather than saving it as `word_t`. There might be some use cases where you still might want to pass on the capability with all of its metadata, but just not tagged (the callee compartment might have other tagged capabilities to re-construct a tagged version of it). 

I am still not sure why we still need new syscalls if the user now has the option to completely manage whether to pass tagged pointers in the IPC buffer and/or fast message registers or not? The only scenario I can think of is when the user disables propagating capabilities in the IPC buffer *and* message registers, but still wants to Read/Write a HW register. Is there any other reason? 


> > I'd like for this thread to be about whether the IPC buffer storage should change it's type.
> 
> If no tagged pointers are passed via the IPC buffer (because that would be passing CHERI pointers via memory), its type stays the same. That's why I replied here.

Reply:
No, we don't agree on that. Maybe Kent does, but I don't. I think it would increase the attack surface too much in practice if we would allow user space to do that. You don't for CHERI-BSD either, why open this gaping hole for seL4?

On the one hand we want minimal changes to the kernel and on the other hand we want something that makes sense and is secure. I think we agree on this, but not on the details.

Allowing user space to pass pointers via IPC (either via registers or the buffer) within the same address space is something that makes sense and wouldn't be totally insecure, we do agree on that. However, it conflicts with minimal changes to the kernel and it also makes things less secure if not all details are handled correctly (either in the implementation or in user space). For instance:

> Retaining the tag or not is better handled by explicitly "clearing" the tag bit rather than saving it as word_t. 

That would make things insecure. You really want deny by default, and only allow explicitly in specific cases when passing tagged pointers through the kernel.

The minimal change for the kernel is supporting just enough that CHERI-enabled user space is possible. For that, one task needs to be able to set another task's registers to tagged pointers. This can be implemented via either:

1) supporting tagged register arguments for some syscalls
2) or supporting tagged pointers in the IPC buffer and getting them that way
3) or not supporting tagged pointer passing altogether and creating them in the kernel.

1 seems a less intrusive of a change than 2. There is no reason to support both. I haven't heard a good argument why 3 isn't sufficient yet, so personally I haven't ruled it out. However, practically it seems simple enough to implement 1 with minimal changes.

And another important reason: We can always add support for passing tagged pointers via user space IPC later (either registers or the buffer). This is a strong hint it's not part of a minimal implementation.

Reply:
> No, we don't agree on that. Maybe Kent does, but I don't. I think it would increase the attack surface too much in practice if we would allow user space to do that. You don't for CHERI-BSD either, why open this gaping hole for seL4?
> 
Please correct me if I am wrong, but we agreed to make it an option and let the user decide its security policy, not to globally prohibit it. CheriBSD simply doesn't have Read/Write UserContext and doesn't off-load all servers and fault handlers to user-space in different address spaces like in seL4. The gaping hole is filled now by allowing the user to globally disable capabilities during IPC, and even if it does allow it, the user can setup a secure CHERI system on top by having non overlapping address spaces for mutually distrusting or untrustworthy threads. 


> On the one hand we want minimal changes to the kernel and on the other hand we want something that makes sense and is secure. I think we agree on this, but not on the details.
> 
> Allowing user space to pass pointers via IPC (either via registers or the buffer) within the same address space is something that makes sense and wouldn't be totally insecure, we do agree on that. However, it conflicts with minimal changes to the kernel and it also makes things less secure if not all details are handled correctly (either in the implementation or in user space). For instance:
>

Yeah we agree on that. I've been trying to meet halfway here by only adding the most necessary changes but also to allow CHERI userspace to be usable (from our side as CHERI developers). Hence this hybrid kernel and other features I stripped out. I think that's minimal enough to make both seL4 and CHERI useable. After all it's like an architectural port and/or big features, so it's expected to big relatively non-trivial.

I am not sure how it could still be less secure if we now give the user the option to completely disable passing tagged pointers during IPC by default? 

> > Retaining the tag or not is better handled by explicitly "clearing" the tag bit rather than saving it as word_t.
> 
> That would make things insecure. You really want deny by default, and only allow explicitly in specific cases when passing tagged pointers through the kernel.
> 
> The minimal change for the kernel is supporting just enough that CHERI-enabled user space is possible. For that, one task needs to be able to set another task's registers to tagged pointers. This can be implemented via either:
> 
> 1. supporting tagged register arguments for some syscalls
> 2. or supporting tagged pointers in the IPC buffer and getting them that way
> 3. or not supporting tagged pointer passing altogether and creating them in the kernel.
> 
> 1 seems a less intrusive of a change than 2. There is no reason to support both. I haven't heard a good argument why 3 isn't sufficient yet, so personally I haven't ruled it out. However, practically it seems simple enough to implement 1 with minimal changes.
> 
> And another important reason: We can always add support for passing tagged pointers via user space IPC later (either registers or the buffer). This is a strong hint it's not part of a minimal implementation.

OK so I guess we still disagree on what minimal is, even though, part of this PR, we have all the solutions the make the user implement the most secure/restrictive CHERI system on top, while allowing CHERI to be usable (from the perceptive of us in the CHERI community). 

I still am happy to add your custom syscalls, for instance, when the user chooses to disable tagged pointers altogether during vanilla seL4 IPC, but I don't think this should prohibit other security policies and threat models I mentioned before and this PR allows the user to have and configure.

Reply:
>3) Follow 2), while trying to be more intentional and use `seL4_Register` in shared code (non-CHERI and CHERI) to differentiate between using integers, pointers, messages registers, HW registers, etc.

I think option 3 should be considered:
- Specializing the seL4_Word to seL4_Register as a typedef is a pattern used elsewhere in libsel4 already,
- It does distinguish that the type of the field is "register" based and does map to register transfers by the kernel,
- The type change doesn't create any ABI break in non-CHERI architectures,
- It shouldn't break existing C client code because all supported architectures have the same original type. 
- It gives a way for existing C code to be ported to CHERI aware while still being able to work with non-CHERI arch targets.

One possible downside is that non-C languages that generate language bindings via parsing the existing C headers may experience a breakage (eg a Rust bindgen approach), but we could ask about this when proposing the seL4_Register addition (it would need a small RFC).

Reply:
> @kent-mcleod Any preference on that change? I am happy to do either

I think my preference would need to be informed by trying to apply user level code to the changes. I know there's code that exists that uses a single 4k page for IPC buffer and some other thread specific data such as static TLS, and maybe it'd be better to keep the total IPC buffer size the same between C64 and A64.

Reply:
> > @kent-mcleod Any preference on that change? I am happy to do either
> 
> I think my preference would need to be informed by trying to apply user level code to the changes. I know there's code that exists that uses a single 4k page for IPC buffer and some other thread specific data such as static TLS, and maybe it'd be better to keep the total IPC buffer size the same between C64 and A64.

Even when we increase the IPC buffer for CHERI, it'd still fit under 4k page. That is, it's only the structure size itself that got changed, but whenever you allocate an IPC buffer, it's still a 4k page as usual. No TLS or API changes were further required for that in sel4test/sel4bench. 

> I think to merge something back upstream there should be acceptance of at least the design approach/goals and an acceptable implementation. I agree that transferring CHERI pointers across different virtual address spaces breaks a bunch of the monotonicity and confinement properties that CHERI enables. 

Yeah I am not disagreeing with that. I thought for this RFC:
1. We concluded that we globally, by default, forbid passing tagged pointers inter-AS.
2. We give the config option to the user to allow that for different threat models, given that they know what they're doing.
3. Even if the user allows 2, it still has the option to have strict CHERI security and avoid pointer aliasing by having non-overlapping address spaces for mutually distrusting (and untrustworthy compartments).

That way, we're making sure both seL4 and CHERI communities make the best use of this core CHERI-seL4 PR. We give the user the option to implement the most strict and/or informed flexible security policy based on their threat models.  Where's the security gap here or the disagreement? What are the blockers? 

> There's no way to use a mark and sweep approach for revoking capabilities if they can move across an address space for example.

There are CHERI approaches to do so and to revoke propagated capabilities, however, this is a temporal safety issue that's beyond the scope of this project and isn't required to still have spatial memory safety. I listed it as future work, and this PR does allow it to be implemented by exporting the relative CHERI PTE bits. CheriBSD didn't have that for quite sometime, so it's totally fine for CHERI-seL4 to offer some spatial memory security and not address *all* CHERI issues we can think of immediately before this RFC/PR gets accepted; given that we know there are solutions for them and we don't forbid them.  As I mentioned, this is just an incremental step to allow sensible CHERI use-cases on top with flexible security policies based on user's threat model.

Reply:
> > 3. Follow 2), while trying to be more intentional and use `seL4_Register` in shared code (non-CHERI and CHERI) to differentiate between using integers, pointers, messages registers, HW registers, etc.
> 
> I think option 3 should be considered:
> 
> * Specializing the seL4_Word to seL4_Register as a typedef is a pattern used elsewhere in libsel4 already,
> * It does distinguish that the type of the field is "register" based and does map to register transfers by the kernel,
> * The type change doesn't create any ABI break in non-CHERI architectures,
> * It shouldn't break existing C client code because all supported architectures have the same original type.
> * It gives a way for existing C code to be ported to CHERI aware while still being able to work with non-CHERI arch targets.
> 
This PR does all of that. 

> One possible downside is that non-C languages that generate language bindings via parsing the existing C headers may experience a breakage (eg a Rust bindgen approach), but we could ask about this when proposing the seL4_Register addition (it would need a small RFC).